File: oth_board.py
"""
Board representation for the game Othello
includes logic for legal moves, and ability to undo moves
(includes move history)
"""

EMPTY = 0
BLACK = 1
WHITE = 2
BORDER = 3

DIRS = [(-1, -1), (-1, 0), (-1, 1), 
        (0, -1), (0, 1),
        (1, -1), (1, 0), (1, 1)]

LETTERS = [chr(x) for x in range(ord('A'), ord('Z')+1)]

SYMBOLS = {
    EMPTY: '.',
    BLACK: 'X',
    WHITE: 'O'}

def opp(color):
    """
    returns opponent of given color
    """
    if color == BLACK:
        return WHITE
    elif color == WHITE:
        return BLACK
    else:
        assert(False)

class OthBoard:

    def __init__(self, size):
        """
        create Othello Board of size <size * size>
        max size of 10 for now
        """
        assert(size > 0 and size < 10)
        # you should only be able to set odd size if you choose to
        assert(not size % 2)

        self.allow_odd_size_boards = True
        self.size = size
        self.komi = 0.5 # for simplification purposes for now
        self.Reset()

    def ChangeSize(self, size):
        """
        change size of board, and also reset it
        """
        if size % 2 and not self.allow_odd_size_boards:
            raise ValueError("size of Othello board must be even.")
        
        self.size = size
        self.Reset()

    def Reset(self):
        """
        reset to empty board, beginning of game
        """
        n = self.size ** 2
        self.board = [EMPTY] * n
        self.move_history = []
        self.current_player = BLACK

        m2 = self.size // 2
        m1 = m2 - 1

        self.Place((m1, m1), WHITE)
        self.Place((m2, m2), WHITE)
        self.Place((m1, m2), BLACK)
        self.Place((m2, m1), BLACK)

    def __str__(self):
        """
          A B C D
        1 . . . . 
        2 . X O . 
        3 . O X .
        4 . . . .
        to play: {BLACK, WHITE}
        """
        size = self.size
        tmp_board = [[EMPTY] * size for _ in range(size)]

        for i in range(size):
            for j in range(size):
                tmp_board[i][j] = self.AccessBoard(i, j)

        lines = ['  ' + ' '.join(LETTERS[0:self.size])]

        # transpose the 2D array so it prints out correctly
        cpboard = [*zip(*tmp_board)]

        for i, row in enumerate(cpboard):
            line = [str(i+1)]
            if self.size > 9 and i < 9:
                line[0] += ' '
            line += [SYMBOLS[x] for x in row]
            lines.append(' '.join(line))

        lines.append("to play: {}".format(SYMBOLS[self.CurrentPlayer()]))
        return '\n'.join(lines)

    def __hash__(self):
        """
        creates hash of the board quickly
        """
        return hash(''.join([SYMBOLS[x] for x in self.board]) + str(self.current_player))

    def PointToStr(self, move):
        """
        for example: (0, 0) -> "A1"
        """
        return LETTERS[move[0]] + str(move[1]+1)

    def StrToPoint(self, _str):
        """
        for example: "A1" -> (0, 0)
        """
        _str = _str.upper()
        return ord(_str[0]) - ord('A'), int(_str[1]) - 1

    def CurrentPlayer(self):
        """
        player who moves next
        """
        return self.current_player

    def CurrentPlayerStr(self):
        """
        string representation of current player (X, O)
        """
        return SYMBOLS[self.current_player]

    def InBounds(self, n):
        """
        if index is within bounds of board (any dimension)
        """
        return 0 <= n < self.size

    def AllPointsBeside(self, p):
        """
        returns array of all points adjacent (including diagonally)
        from point p
        """
        if isinstance(p, str):
            p = self.StrToPoint(p)

        x, y = p
        points = []
        for dx, dy in DIRS:
            if self.InBounds(x+dx) and self.InBounds(y+dy):
                points.append((x+dx, y+dy))
        return points

    def AccessBoard(self, x, y):
        """
        returns the color of the board at point x, y
        """
        sz = self.size
        idc = sz * x + y
        return self.board[idc]

    def SetBoard(self, x, y, color):
        """
        sets the color of the board at point x, y
        """
        sz = self.size
        idc = sz * x + y
        self.board[idc] = color

    def GetCaptures(self, move):
        """
        returns all points captured by this move
        """
        our_color = self.CurrentPlayer()

        x, y = move

        if not self.InBounds(x) or not self.InBounds(y):
            return []

        if self.AccessBoard(x, y) != EMPTY:
            return []

        all_captures = []

        legal = False
        # check in all 8 directions (othello works diagonally)
        for dx, dy in DIRS:
            cx, cy = move
            cx, cy = cx + dx, cy + dy

            seen_opp = False
            tmp_captures = []

            # cx tracks where we are checking
            while self.InBounds(cx) and self.InBounds(cy):

                if self.AccessBoard(cx, cy) == opp(our_color):
                    tmp_captures.append((cx, cy))
                    seen_opp = True

                elif self.AccessBoard(cx, cy) == EMPTY:
                    # illegal line of captures
                    tmp_captures = []
                    break

                elif self.AccessBoard(cx, cy) == our_color and seen_opp:
                    # legal line of capture
                    all_captures += tmp_captures
                    tmp_captures = []
                    break

                cx += dx
                cy += dy

        return all_captures

    def NumCaptured(self, move):
        """
        :param move: the move of a player
        :return: returns an int of the number of pieces captured by a move
        """
        if isinstance(move, str):
            move = self.StrToPoint(move)

        return len(self.GetCaptures(move))

    def EvaluateMove(self, move):
        """
        :param move: The move taken by the agent (needs to be (x, y) tuple form)
        :return: retuns a tuple of two floats based on the desirability of the current state and move
        """
        move = self.StrToPoint(move)

        self.Play(move)
        legal_moves = self.GetLegalMoves()
        player_count = 0
        opponent_count = 0
        move_score = 0.0
        for i in range(self.size):
            for j in range(self.size):
                if self.AccessBoard(i, j) == self.current_player:
                    player_count += 1
                elif self.AccessBoard(i, j) != 0:
                    opponent_count += 1

        # Move Score update on Mobility
        if len(self.move_history) != 0:
            move_score += len(legal_moves) / len(self.move_history)

        self.Undo()

        # Get the difference in the total discs captured.
        captured = self.NumCaptured(move) * len(self.move_history)
        difference = (player_count + captured * 2 + 1 - opponent_count) * len(self.move_history)

        # Start the original action value with the negative number of disc captures
        move_score += float(captured) + difference

        # Check if move creates corner state --> better (checked at all times)
        limit = self.size - 1
        corners = [(0, 0), (0, limit), (limit, 0), (limit, limit)]

        # Check if the current move is a corner play, or if the current state can allow for a corner play
        # Update both values
        for corner in corners:
            if move == corner:
                move_score += -300.0

        return move_score

    # Evaulate the goodness of a state for depth limited AB
    def EvaluateState(self):

        # Check mobility of the state
        legalMoves = self.GetLegalMoves()
        stateScore = -(len(legalMoves))

        # Check if a corner is available
        limit = self.size - 1
        corners = [(0, 0), (0, limit), (limit, 0), (limit, limit)]

        # Check if the current move is a corner play, or if the current state can allow for a corner play
        # Update both values
        for corner in corners:
            if corner in legalMoves:
                stateScore += -300.0

        return stateScore

    def IsLegal(self, move):
        """
        tells us if move is legal
        """
        if self.GetCaptures(move):
            return True
        return False

    def Play(self, p):
        """
        play at point p 
        """

        if isinstance(p, str):
            if p.lower() == 'pass':
                raise RuntimeError("passing during AB search, incorrect")
                self.current_player = opp(self.current_player)
                return True
            p = self.StrToPoint(p)
        
        captures = self.GetCaptures(p)
        if not captures:
            return False
        
        self.SetBoard(p[0], p[1], self.current_player)
        for x, y in captures:
            self.SetBoard(x, y, self.current_player)

        self.current_player = opp(self.current_player)
        self.move_history.append((p, captures))

        return True

    def Undo(self):
        """
        undo last move on the stack, and uncapture all captured pieces
        set captured to current player
        """
        p, captures = self.move_history.pop()
        x, y = p
        self.SetBoard(x, y, EMPTY)
        for x, y in captures:
            self.SetBoard(x, y, self.CurrentPlayer())
        self.current_player = opp(self.current_player)

    def Terminal(self):
        """
        return True if the game is over for the current state
        False otherwise
        """
        if self.GetLegalMoves():
            return False

        self.current_player = opp(self.current_player)
        if self.GetLegalMoves():
            self.current_player = opp(self.current_player)
            return False
        
        self.current_player = opp(self.current_player)
        return True

    def Winner(self):
        """
        return Winner, Score of the game
        """
        bcount = 0
        wcount = 0
        for i in range(self.size):
            for j in range(self.size):
                if self.AccessBoard(i, j) == BLACK:
                    bcount += 1
                elif self.AccessBoard(i, j) == WHITE:
                    wcount += 1

        score = bcount - wcount + self.komi # komi to make sure we don't tie for now
        if score > 0:
            return BLACK, score
        elif score < 0:
            return WHITE, score
        else:
            raise ValueError("TIE OCCURRED WITH KOMI")
            return EMPTY, score

    def Hash(self):
        """
        return hash of the board's state (not including history)
        """
        return hash(self)

    def Place(self, point, color):
        """
        place a piece without checking legality.
        replaces currently placed pieces
        """
        assert(color in [BLACK, WHITE, EMPTY])
        r, c = point
        self.SetBoard(r, c, color)

    def GetLegalMoves(self):
        moves = []
        for i in range(self.size):
            for j in range(self.size):
                if self.IsLegal((i, j)):
                    moves.append((i, j))
        return [self.PointToStr(m) for m in moves]

File: cpp/oth_board.cpp
/*
Board representation for the game Othello
includes logic for legal moves, and ability to undo moves
(includes move history)
*/

#include "oth_board.h"

#include <iostream>
#include <sstream>

OthBoard::OthBoard(int size)
{   
    // make sure its not too big
    assert(size > 0 && size < 10);

    m_allowOddSizedBoards = true;
    m_size = size;
    m_komi = 0.5f;
    Reset();
    InitDirs();
}

OthBoard::~OthBoard()
{ }

void OthBoard::InitDirs()
{
    m_dirs.push_back(p(-1, -1));
    m_dirs.push_back(p(-1, 0));
    m_dirs.push_back(p(-1, 1));
    m_dirs.push_back(p(0, -1));

    m_dirs.push_back(p(0, 1));
    m_dirs.push_back(p(1, -1));
    m_dirs.push_back(p(1, 0));
    m_dirs.push_back(p(1, 1));

    m_dirs.push_back(p(0, 0));
}

void OthBoard::Reset()
{
    int n = m_size * m_size;
    m_board = othColorList(n, EMPTY);
    m_history = othHistory();
    m_currentPlayer = BLACK;

    int m2 = m_size / 2;
    int m1 = m2 - 1;

    Place(p(m1, m1), WHITE);
    Place(p(m2, m2), WHITE);
    Place(p(m1, m2), BLACK);
    Place(p(m2, m1), BLACK);
}

std::string ColorToString(othColor c)
{
    if (c == WHITE)
        return std::string("O");
    
    else if (c == BLACK)
        return std::string("X");
    
    else if (c == EMPTY)
        return std::string(" ");

    return std::string();
}

std::string OthBoard::Str()
{
    std::stringstream ss;
    ss << " ";
    for (int i = 0; i < m_size; i++)
    {
        ss << " " << ((char) (i + 'A'));
    }
    ss << std::endl;

    for (int i = 0; i < m_size; i++)
    {
        ss << i+1;
        for (int j = 0; j < m_size; j++)
        {
            ss << " " << ColorToString(AccessBoard(p(j, i)));
        }
        ss << std::endl;
    }
    ss << "to play: " << ColorToString(CurrentPlayer());
    return ss.str();
}

std::string OthBoard::PointToString(othPoint point) const
{
    return std::string{(char)(point.first + 'A'), (char)(point.second + '1')};
}

othPoint OthBoard::StrToPoint(std::string str) const
{
    int x = str[0] - 'A';
    int y = str[1] - '1';
    return p(x, y);
}

void OthBoard::ChangeSize(int size)
{
    m_size = size;
    Reset();
}

othColor OthBoard::CurrentPlayer() const
{
    return m_currentPlayer;
}

std::string OthBoard::CurrentPlayerStr() const
{
    if (m_currentPlayer == BLACK)
        return std::string("black");
    
    if (m_currentPlayer == WHITE)
        return std::string("white");
    
    // current player must be BLACK or WHITE
    assert(false);
}

bool OthBoard::InBounds(othPoint point) const
{
    return (point.first >= 0 && point.first < m_size)
        && (point.second >= 0 && point.second < m_size);
}

othPointList OthBoard::AllPointsBeside(othPoint point) const
{
    int x = point.first, y = point.second;
    othPointList ret;
    for (auto const& dp : m_dirs)
    {
        int dx = dp.first, dy = dp.second;
        othPoint pt = p(x+dx, y+dy);

        if (InBounds(pt))
        {
            ret.push_back(pt);
        }
    }
    return ret;
}

othColor OthBoard::AccessBoard(othPoint pt) const
{
    int sz = m_size;
    int idx = sz * pt.first + pt.second;
    return m_board[idx];
}

void OthBoard::SetBoard(othPoint pt, othColor c)
{
    int sz = m_size;
    int idx = sz * pt.first + pt.second;
    m_board[idx] = c;
}

othPointList OthBoard::GetCaptures(othPoint pt) const
{
    othColor our_color = CurrentPlayer();
    int x = pt.first, y = pt.second;

    if (!InBounds(pt))
    {
        return othPointList();
    }

    if (AccessBoard(pt) != EMPTY)
    {
        return othPointList();
    }

    othPointList all_captures = othPointList();

    bool legal_flag = false;
    // check in all 8 directions (othello works diagonally too)
    for (auto const& dpt : m_dirs)
    {
        int dx = dpt.first, dy = dpt.second;
        int cx = x+dx, cy = y+dy;

        bool seen_opp = false;
        othPointList tmp_captures = othPointList();

        othPoint cpt = p(cx, cy);
        while (InBounds(cpt))
        {
            if (AccessBoard(cpt) == opp(our_color))
            {
                tmp_captures.push_back(cpt);
                seen_opp = true;
            }

            else if (AccessBoard(cpt) == EMPTY)
            {
                tmp_captures.clear();
                break;
            }

            else if (AccessBoard(cpt) == our_color && seen_opp)
            {
                all_captures.insert(all_captures.end(), 
                                    tmp_captures.begin(), 
                                    tmp_captures.end());
                tmp_captures.clear();
                break;
            }

            cx += dx;
            cy += dy;
            cpt = p(cx, cy);
        }
    }
    return all_captures;
}

bool OthBoard::IsLegal(othPoint pt) const
{
    if (!GetCaptures(pt).empty())
    {
        return true;
    }
    return false;
}

bool OthBoard::Play(othPoint pt)
{
    if (is_pass(pt)) {
        // shouldn't be able to pass during AB search
        assert(false);
        m_currentPlayer = opp(m_currentPlayer);
    }

    auto captures = GetCaptures(pt);
    if (captures.empty())
        return false;

    SetBoard(pt, CurrentPlayer());
    for (auto const& cpt : captures)
        SetBoard(p(cpt.first, cpt.second), CurrentPlayer());

    m_currentPlayer = opp(CurrentPlayer());
    // todo: maybe push pointers with new here, to avoid copies
    m_history.push_back(make_pair(pt, captures));

    return true;
}

void OthBoard::Undo()
{
    othPoint pt = m_history.back().first;
    othPointList captures = m_history.back().second;
    m_history.pop_back();

    SetBoard(pt, EMPTY);
    for (auto const& cpt : captures)
        SetBoard(cpt, CurrentPlayer());

    m_currentPlayer = opp(m_currentPlayer);
}

bool OthBoard::Terminal()
{
    if (GetLegalMoves().empty())
        return false;

    m_currentPlayer = opp(m_currentPlayer);
    if (GetLegalMoves().empty())
    {
        m_currentPlayer = opp(m_currentPlayer);
        return false;
    }

    m_currentPlayer = opp(m_currentPlayer);
    return true;
}

std::pair<othColor,float> OthBoard::Winner() const
{
    int bcount = 0, wcount = 0;
    
    for (int i = 0; i < m_size; i++)
    {
        for (int j = 0; j < m_size; j++)
        {
            othColor c = AccessBoard(p(i, j));
            if (c == BLACK)
            {
                bcount++;
            }
            else if (c == WHITE)
            {
                wcount++;
            }
        }
    }

    float score = bcount - wcount + m_komi; // komi to make sure we don't tie for now
    if (score > 0.0)
    {
        return std::make_pair(BLACK, score);
    }
    else if (score < 0.0) 
    {
        return std::make_pair(WHITE, score);
    }
    else 
    {
        // score should not be 0 with komi
        assert(false);
        return std::make_pair(EMPTY, score);
    }
}

uint64_t OthBoard::Hash() const
{
    return 1;
}

void OthBoard::Place(othPoint pt, othColor c)
{
    assert(c == BLACK || c == WHITE || c == EMPTY);
    int x = pt.first, y = pt.second;
    SetBoard(p(x, y), c);
}

othPointList OthBoard::GetLegalMoves()
{
    othPointList moves;
    for (int i = 0; i < m_size; i++)
    {
        for (int j = 0; j < m_size; j++)
        {
            othPoint pt = p(i, j);
            if (IsLegal(pt))
            {
                moves.push_back(pt);
            }
        }
    }
    return moves;
}
